python/pyspark/sql/session.py

```diff
@@ -697,12 +712,17 @@ def prepare(obj):
                 verify_func(obj)
                 return obj,
         else:
+            no_need_to_prepare = True
             prepare = lambda obj: obj
 
         if isinstance(data, RDD):
-            rdd, schema = self._createFromRDD(data.map(prepare), schema, samplingRatio)
+            rdd, schema = self._createFromRDD(
+                data if no_need_to_prepare else data.map(prepare),
```

Comment:
I think you could just say `data if not verifySchema else ...` and remove `no_need_to_prepare`

Reply:
There are cases when verifySchema is True and the `prepare` function is the identity function.

That's why we need the `no_need_to_prepare` flag.